app/models/organization_table_info.rb

```diff
@@ -0,0 +1,165 @@
+class OrganizationTableInfo
+  attr_accessor :organizations_counts, :total_counts
+
+  def initialize(org_stats, total_counts)
+    @organizations_counts = org_stats.kind_of?(Array) ? org_stats : [ org_stats ]
+    @total_counts = total_counts
+  end
+
+  def self.top_stats(top: 0, organization: nil, weeks: 5)
+    # Retrieve the results count, payloads count and organization info for the top most consuming organizations
+    org_payloads_and_results_count = Organization
+      .select('
+        organizations.id, organizations.api_id, organizations.name, organizations.display_name,
+        SUM(test_payloads.results_count) AS test_results_count, COUNT(test_payloads.id) as test_payloads_count
+      ')
+      .joins('
+        LEFT OUTER JOIN "projects" ON "projects"."organization_id" = "organizations"."id"
+        LEFT OUTER JOIN "project_versions" ON "project_versions"."project_id" = "projects"."id"
+        LEFT OUTER JOIN"test_payloads" ON "test_payloads"."project_version_id" = "project_versions"."id"
+      ')
+      .group('organizations.id')
+      .order('test_results_count DESC NULLS LAST')
+
+    # Retrieve the results trends for organizations
+    org_results_trends_count = Organization
+      .select("
+        organizations.id,
+        SUM(test_payloads.results_count) AS test_results_count,
+        date_trunc('week', test_payloads.processed_at) AS trend_date
+      ")
+      .joins('
+        LEFT OUTER JOIN "projects" ON "projects"."organization_id" = "organizations"."id"
+        LEFT OUTER JOIN "project_versions" ON "project_versions"."project_id" = "projects"."id"
+        LEFT OUTER JOIN"test_payloads" ON "test_payloads"."project_version_id" = "project_versions"."id"
+      ')
+      .group("organizations.id, date_trunc('week', test_payloads.processed_at)")
+      .order('organizations.id, trend_date ASC')
+      .where("date_trunc('week', test_payloads.processed_at) >= ?", weeks.weeks.ago.at_beginning_of_week)
+
+    # Add where clause to scope to specific organization
+    if organization
+      org_payloads_and_results_count = org_payloads_and_results_count.where('organizations.id = ?', organization.id)
+      org_results_trends_count = org_results_trends_count.where('organizations.id = ?', organization.id)
+    end
+
+    # Add a limit to do the top n
+    if top > 0
+      org_payloads_and_results_count = org_payloads_and_results_count.limit(top)
+    end
+
+    # Collect the ids
+    org_ids = org_payloads_and_results_count.collect(&:id)
```

Comment:
`org_payloads_and_results_count` is an Active Record relation, not an array of results. Up until this line, the query described by the relation has not yet been executed. When you call `collect` on it, it executes the query and proxies the `collect` method to the result array. Later, at line 95, you again call `each` on the relation, which it also proxies to the result array. It doesn't execute the query twice because the relation hasn't changed so it probably caches the results.

It works but I personally prefer to explicitly transform the relation into a result array:

```
org_payloads_and_results_count_rel = Organization...
org_payloads_and_results_count = org_payloads_and_results_count_rel.to_a
org_ids = org_payloads_and_results_count.collect(&:id)
org_payloads_and_results_count.each do |payloads_and_results_count|
  ...
end
```

It makes it clearer what is an Active Record relation and what is an actual result.

A similar refactoring could be done for the `org_results_trends_count` relation:

```
org_ids.each do |org_id|
  org_results_trends = org_results_trends_count.select{ |org_results_trend| org_results_trend.id == org_id }
  ...
end
```

`select` is called multiple times on the relation, presumably returning the same cached results. But it would be clearer to explicitly retrieve those results before the parent `each`.